### PR TITLE
Discard records when the buffer is full

### PR DESCRIPTION
**Pull request by aviflax**
This way an application can decide what to do if/when
the buffer is full — some apps may prefer to just discard
records, or they might buffer them to disk or something.

Refs [#3134](https://timehop.clubhousehq.com/story/3134)

@bdotdub or @brunodecarvalho please review.

### CODE REVIEW

File: batchproducer/batchproducer.go [hunk 1]
```diff
@@ -224,6 +238,14 @@ func (b *batchProducer) SetStatInterval(si time.Duration) error {
...
+func (b *batchProducer) SetProblemPolicy(pp ProblemPolicy) error {
+	if b.isRunning() {
+		return errors.New("Cannot set problem policy while Producer is running.")
+	}
+	b.problemPolicy = pp
+	return nil
+}
+
 func (b *batchProducer) setRunning(running bool) {
 	b.runningMu.Lock()
```
Comment (biasedbit) [inline batchproducer/batchproducer.go:236]: I'd use an atomic bool here since we check for running quite often.

File: batchproducer/batchproducer.go [hunk 2]
```diff
@@ -36,6 +36,9 @@ type Producer interface {
 	// when this interval elapses. There’s no guarantee, however, since the main goroutine is
 	// used to send the stats and therefore there may be some skew.
 	SetStatInterval(time.Duration) error
+
+	// SetProblemPolicy sets the ProblemPolicy.
+	SetProblemPolicy(ProblemPolicy) error
```
Comment (biasedbit) [inline batchproducer/batchproducer.go]: I think I'd rather see a `Config` object, passed on initialization, than a bunch of pre-run setters (i.e. that stop working as soon as you start the component.)

File: batchproducer/batchproducer.go (hunk 1 above)
Comment (aviflax) [inline batchproducer/batchproducer.go:236]: I’ll look into that, thanks!

File: batchproducer/batchproducer.go [hunk 3]
```diff
@@ -260,7 +282,15 @@ func (b *batchProducer) sendBatch() {
 		b.consecutiveErrors++
 		b.currentStat.KinesisErrorsSinceLastStat++
 		b.logger.Printf("Error occurred when sending PutRecords request to Kinesis stream %v: %v", b.streamName, err)
-		b.returnRecordsToBuffer(records)
+
+		if b.consecutiveErrors >= 5 && b.isBufferFullOrNearlyFull() {
+			// In order to prevent Add from hanging indefinitely, we start dropping records
+			b.logger.Printf("DROPPING %v records because buffer is full or nearly full and there have been %v consecutive errors from Kinesis", len(records), b.consecutiveErrors)
+		} else {
+			b.logger.Printf("Returning %v records to buffer (%v consecutive errors)", len(records), b.consecutiveErrors)
```
Comment (biasedbit) [inline batchproducer/batchproducer.go:276]: Why aren't we using our logger like we do on every other TH go project?

File: batchproducer/batchproducer.go (hunk 2 above)
Comment (aviflax) [inline batchproducer/batchproducer.go]: Good thinking, I agree. Will refactor. Thanks!

File: batchproducer/batchproducer.go (hunk 3 above)
Comment (aviflax) [inline batchproducer/batchproducer.go:276]: This is not a TH go project; this is an open-source project by sendgrid and this repo is a fork. I used the existing conventions of the repo because I plan to contribute this addition back to the original project.

File: batchproducer/batchproducer.go (hunk 2 above)
Comment (aviflax) [inline batchproducer/batchproducer.go]: @brunodecarvalho please check out 8b80fc9. Thanks!

File: batchproducer/batchproducer.go (hunk 1 above)
Comment (aviflax) [inline batchproducer/batchproducer.go:236]: I’m gonna do this in a separate PR; it’s not actually related to this branch.

**Comment by aviflax** [comment]
@brunodecarvalho please review 8b80fc9
